halo2-base/README.md

```diff
@@ -0,0 +1,498 @@
+            let mut break_point = vec![];
+            let mut gate_index = 0;
+            let mut row_offset = 0;
+            for ctx in threads {
+                let mut basic_gate = config.basic_gates[phase]
+                        .get(gate_index)
+                        .unwrap_or_else(|| panic!("NOT ENOUGH ADVICE COLUMNS IN PHASE {phase}. Perhaps blinding factors were not taken into account. The max non-poisoned rows is {max_rows}"));
+                assert_eq!(ctx.selector.len(), ctx.advice.len());
+
+                for (i, (advice, &q)) in ctx.advice.iter().zip(ctx.selector.iter()).enumerate() {
+                    let column = basic_gate.value;
+                    let value = if use_unknown { Value::unknown() } else { Value::known(advice) };
+                    #[cfg(feature = "halo2-axiom")]
+                    let cell = *region.assign_advice(column, row_offset, value).cell();
+                    #[cfg(not(feature = "halo2-axiom"))]
+                    let cell = region
+                        .assign_advice(|| "", column, row_offset, || value.map(|v| *v))
+                        .unwrap()
+                        .cell();
+                    assigned_advices.insert((ctx.context_id, i), (cell, row_offset));
+            ...
+                    
+```
+In the case a breakpoint falls on the overlap between two gates (such as chained addition of two values) the cells the breakpoint falls on must be copied to the next column and a new equality constraint enforced between the value of the cell in the old column and the copied cell in the new column. This prevents the circuit from being undersconstratined and preserves the equality constraint from the overlapping gates.
+```rust ignore
+if (q && row_offset + 4 > max_rows) || row_offset >= max_rows - 1 {
+    break_point.push(row_offset);
+    row_offset = 0;
+    gate_index += 1;
+
+// when there is a break point, because we may have two gates that overlap at the current cell, we must copy the current cell to the next column for safety
+    basic_gate = config.basic_gates[phase]
+                .get(gate_index)
+                .unwrap_or_else(|| panic!("NOT ENOUGH ADVICE COLUMNS IN PHASE {phase}. Perhaps blinding factors were not taken into account. The max non-poisoned rows is {max_rows}"));
+    let column = basic_gate.value;
+
+    #[cfg(feature = "halo2-axiom")]
+    {
+        let ncell = region.assign_advice(column, row_offset, value);
+        region.constrain_equal(ncell.cell(), &cell);
+    }
+    #[cfg(not(feature = "halo2-axiom"))]
+    {
+        let ncell = region
+                    .assign_advice(|| "", column, row_offset, || value.map(|v| *v))
+                    .unwrap()
+                    .cell();
+        region.constrain_equal(ncell, cell).unwrap();
+    }
+}
+
+```
+
+If `witness_gen_only` flag is set witness generation is only performed. `sub_synthesize` calls `assign_threads_in()` to assign threads directly.
```

Comment:
change to "only witness generation is performed"

emphasize that if `witness_gen_only = true`, then no copy constraints or selectors are considered.